src/zenml/orchestrators/step_launcher.py

```diff
@@ -131,11 +133,79 @@ def __init__(
         self._stack = Stack.from_model(deployment.stack)
         self._step_name = step.spec.pipeline_parameter_name
 
+        # Internal properties and methods
+        self._step_run: Optional[StepRunResponse] = None
+        self._setup_signal_handlers()
+
+    def _setup_signal_handlers(self) -> None:
+        """Set up signal handlers for graceful shutdown."""
+
+        def signal_handler(signum: int, frame: Any) -> None:
+            """Handle shutdown signals gracefully.
+
+            Args:
+                signum: The signal number.
+                frame: The frame of the signal handler.
+
+            Raises:
+                RunStoppedException: If the pipeline run is stopped by the user.
+                RunInterruptedException: If the execution is interrupted for any
+                    other reason.
+            """
+            logger.info(
+                f"Received signal shutdown {signum}. Requesting shutdown "
+                f"for step '{self._step_name}'..."
+            )
+
+            try:
+                client = Client()
+                pipeline_run = None
+
+                if self._step_run:
+                    pipeline_run = client.get_pipeline_run(
+                        self._step_run.pipeline_run_id
+                    )
+                else:
+                    pipeline_run_list = client.list_pipeline_runs(
+                        deployment_id=self._deployment.id,
+                        orchestrator_run_id=self._orchestrator_run_id,
+                        hydrate=False,
+                    )
+                    if pipeline_run_list.items:
+                        pipeline_run = pipeline_run_list.items[0]
+
+                if pipeline_run and pipeline_run.status in [
+                    ExecutionStatus.STOPPING,
+                    ExecutionStatus.STOPPED,
+                ]:
```

Comment:
shouldn't you publish this status update regardless of the pipeline run status ? What if one step gets killed without anyone stopping the parent pipeline run?

Reply:
This is the special case. If your scenario happens, we raise a `RunInterruptedException` which ends up tagging the step as `Failed`.